tests/pos/typeclass-encoding.scala

```diff
@@ -0,0 +1,91 @@
+/** A possible type class encoding for
+
+      trait SemiGroup {
+        def add(that: This): This
+      }
+
+      trait Monoid extends SemiGroup {
+        static def unit: This
+      }
+
+      extend Int : Monoid {
+        def add(that: Int) = this + that
+        static def unit = 0
+      }
+
+      extend String : Monoid {
+        def add(that: Int) = this ++ that
+        static def unit = ""
+      }
+
+      def sum[T: Monoid](xs: List[T]): T =
+        (instance[T, Monoid].unit /: xs)(_ `add` _)
+
+*/
+object runtime {
+
+  trait TypeClass {
+    type This
+    type StaticPart[This]
+  }
+
+  trait Implementation[From] {
+    type This = From
+    type Implemented <: TypeClass
+    def inject(x: From): Implemented { type This = From }
+  }
+
+  class CompanionOf[T] { type StaticPart[_] }
+
+  def instance[From, To <: TypeClass](
+      implicit ev1: Implementation[From] { type Implemented = To },
+      ev2: CompanionOf[To]): Implementation[From] { type Implemented = To } & ev2.StaticPart[From] =
+    ev1.asInstanceOf  // can we avoid the cast?
```

Comment:
why not swap `ev1` and `ev2` and make `ev1` (now second argument) have type `Implementation[From] { type Implemented = To } & ev2.StaticPart[From]`? Let's ignore inference issues for a second.*

It seems all that could be replaced by replacing, e.g., `MonoidStatic[T] with Implementation[T]` with
```scala
class MonoidDict[T] extends Implementation[T] {
  def unit: T
   def inject($this: T) = new Monoid {
      type This = T
      def add(that: T): T
   }
}
```

In fact, at that point it's clearer that MonoidDict instances encode differently the same info as

```scala
trait MonoidDict[T] extends Implementation[T] {
  def unit: T
  def add($this: T, that: T): T
}
```

* If inference has problems with that, maybe we want to get curried implicits in first, but I'd first try simplifying the encoding further.

Reply:
This is by no means the last word in encodings. It's a first attempt to see what it would take. We should experiment with lots of other schemes.